GUIDELINES.md

```diff
@@ -0,0 +1,994 @@
+```
+
+But consuming or configuring a sub-classed dependency won't work (sweetest will complain that the type was not found in the module configuration):
+
+```kotlin
+private val instance by dependency<FakeBackendGateway>() // <-- doesn't work
+
+override fun configure() = super.configure()
+    .requireReal<FakeBackendGateway>() // <-- doesn't work 
+```
+
+In cases of abstract types (abstract classes or interfaces) or when there are different types of an inheritance hierarchy under test you have to declare the top-most level type only in the module configuration! Never declare multiple types of the same type hierarchy because sweetest currently has troubles picking the right type (unfortunately the picked type can be indeterministic leading to hard to debug failing tests, but this will be fixed in one of the upcoming releases)!
+
+##### Lazy-initialized approach
+
+The most preferred workaround for now looks like this:
+
+```kotlin
+private val _instance by dependency<BackendGateway>()
+private val instance get() = _instance as FakeBackendGateway
+
+override fun configure() = super.configure()
+    .offerMockRequired<BackendGateway> { FakeBackendGateway() }
+```
+
+It's important to have `<BackendGateway>` in `dependency` and `offerMockRequired` unchanged, because that directs sweetest to the right dependency configuration.
+
+The solution is obviously clunky, but it works in all cases as it preserves the lazy behavior of dependency initializations. The lazy behavior helps in situations where initialization code in production classes get in the way during the initialization of the test.
+
+##### Immediate initialization approach
+
+ ```kotlin
+private val instance = FakeBackendGateway()
+
+override fun configure() = super.configure()
+    .offerMockRequired<BackendGateway> { instance }
+```
+
+This is easier but requires the class (in this case `FakeBackendGateway`) being compatible with it being initialized that early during creation of the steps classes. Feel free to use that approach when suitable.
+
+#### Special case: spy
+
+The approach described above is also of value if you need to create a spy on a class.
+
+**Approach A:**
+
+```kotlin
+private val instance = spy(FakeBackendGateway())
+
+override fun configure() = super.configure()
+    .offerMockRequired<BackendGateway> { instance }
+```
+
+**Approach B:**
+
+But it's also possible to create Mockito spies like so:
+
+```kotlin
+private val instance by dependency<AuthManager>()
+
+override fun configure() = super.configure()
+    .requireSpy<AuthManager>()
+```
+
+The downside using `requireSpy` is that you have no control over the creation of the underlying class that is spied on: an `AuthManager` will be created and wrapped by a spy.
+
+In contrast to that, for the `FakeBackendGateway` we _need_ to use **approach A** as we need more control over the spy creation (we don't need a spy on `BackendGateway` but rather on `FakeBackendGateway`).
+
+### Module testing configuration
```

Comment:
hm, i kept the mentions in the introduction on a surface level intentionally and linked the reference ("for more info see..."). where the intro is here to give a broad understanding, the reference is going into detail. your feedback makes me question whether that goal was really achieved now 🤔 can you probably explain a little more?

Reply:
ok I'll try again :)
Coming from the reading it top to bottom, as a whole, I felt like some information is repeating (same in intro and ref, maybe different words) and some is adding more depth in the reference but also some things are _better_ explained in the introduction.

As I understand you'd prefer to be the reference more stand-alone and the introduction is like an add-on.. 
I thought, thats why I also mentioned it before, as a whole document maybe a different structure would be beneficial (e.g. theory/basics, reference, tips, example/tutorial/intro) and to merge the spread information.

Another thing could be to just shorten the introduction to an example with one sentence to something like steps and always link only to the reference for more (there the information is merged on this topic), since the basic names and stuff have to be read anyways..

Hope I could make my thoughts/intentions clear.. it is really just something that came to mind and don't get me wrong it is just thinking out loud :)